### ISSUE TITLE
Ops with implicit dim order conversion fail at runtime

### ISSUE THREAD

**Issue by GregoryComer** [pytorch/executorch#11086]
### 🐛 Describe the bug

Certain ops in pytorch can implicitly convert their inputs to contiguous format. Linear, for example, seems to do this. These dim order conversions are reflected in the node metadata (and thus ET dim order), but not in the graph. Our kernels don't include the input dim order conversion and thus fail at runtime.

A slightly contrived example is as follows, though I expect that this is reproable with any op with an ATen kernel that includes an implicit conversion to contiguous, of which there are quite a few.

```py
import torch

class Module(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(10, 10)

    def forward(self, x):
        return self.linear(x)

inputs = (torch.randn(1,2,3,10).to(memory_format=torch.channels_last),)
module = Module()

ep = torch.export.export(module, inputs)
from executorch.exir import to_edge_transform_and_lower

et_program = to_edge_transform_and_lower(ep).to_executorch()
from executorch.extension.pybindings.portable_lib import _load_for_executorch_from_buffer

et_model = _load_for_executorch_from_buffer(et_program.buffer)
```

Output:
```
[tensor_util_portable.cpp:132] Check failed (all_contiguous || all_channels_last): 2 input tensors have different dim orders
[op_expand_copy.cpp:89] Check failed (tensors_have_same_dim_order(self, out)): 
[method.cpp:1322] KernelCall failed at instruction 0:1 in operator aten::expand_copy.out: 0x12
[method.cpp:1328] arg 0 with type id 1
[method.cpp:1328] arg 1 with type id 8
[method.cpp:1328] arg 2 with type id 5
[method.cpp:1328] arg 3 with type id 1
[method.cpp:1328] arg 4 with type id 1
```
Note that the failure actually occurs in expand_copy due to the weird decomp in this case.

### Versions

N/A

**Comment by digantdesai** [comment]
`print(module.forward(*inputs).dim_order())` what does this print?

**Comment by GregoryComer** [comment]
> `print(module.forward(*inputs).dim_order())` what does this print?

It prints `(0, 1, 2, 3)`.

**Comment by digantdesai** [comment]
> These dim order conversions are reflected in the node metadata (and thus ET dim order), but not in the graph.

Seems like eager mode will work OK. And some delegate trying to match the graph will also work OK. Portable Linear with NHWC input may not. So the action here is to insert an explicit convert when running into such ops? How about decomposition handling this at least for Linear?

**Comment by Lawson-Darrow** [comment]
This still reproduces on executorch 1.3.1 (torch 2.12, CPU). A channels_last input through `nn.Linear` fails at runtime in the portable `op_expand_copy` on `tensors_have_same_dim_order(self, out)`, KernelCall 0x12. The node metadata and specs show a dim-order change to contiguous, but no explicit dim-order conversion is materialized before the decomposed `expand_copy`, so the kernel receives mismatched dim orders. The `expand_copy` invariant itself looks correct, so this reads as a lowering gap rather than a kernel bug.

Happy to take this. Should the fix be an EXIR lowering pass that inserts the dim-order conversion after the decomposition and before out-var and memory planning? I would rather match the intended design before I build.

**Comment by akashjainn** [comment]
Still hitting this on main. Smaller repro that doesn't involve Linear or conv - mean/sum reduce to contiguous internally, so the other operand of the add is still channels_last:

```python
import torch
from torch.export import export
from executorch.exir import to_edge_transform_and_lower
from executorch.extension.pybindings.portable_lib import _load_for_executorch_from_buffer

class M(torch.nn.Module):
    def forward(self, t):
        return t + t.mean(dim=[2, 3], keepdim=True)

m = M().eval().to(memory_format=torch.channels_last)
x = torch.randn(1, 4, 6, 6).to(memory_format=torch.channels_last)
prog = to_edge_transform_and_lower(export(m, (x,), strict=True), partitioner=[]).to_executorch()
_load_for_executorch_from_buffer(prog.buffer).run_method("forward", (x,))
```

```
Check failed (all_contiguous || all_channels_last): 3 input tensors have different dim orders
KernelCall failed at instruction 0:2 in operator aten::add.out: 0x12
```

Same thing with mul/sub/div. It's not specific to add - 106 files under kernels/portable/cpu call tensors_have_same_dim_order.

I did try extending LegalizePortableDimOrderPass to insert a _to_dim_order_copy on whichever operand disagrees, and both obvious versions are wrong, so flagging that before someone else burns time on it:

- converge the operands on the op's output dim order: runs fine, silently wrong values for add/mul/sub
- converge on contiguous: values are right, but it adds a second copy in the expand_copy path that test_channels_last_linear_inserts_copy_before_expand_copy counts, and two cases in test_issue_11086_runtime_regressions start failing with 0x12

Looks like the op's own output keeps its original dim order in metadata either way, so the mismatch just moves downstream instead of going away. Probably needs the new order propagated to the output spec, which is more than a local change in that pass, so I stopped there.

Not working on this, just leaving the repro.
